packages/mds-compliance-engine/engine/helpers.ts

```diff
@@ -245,93 +212,29 @@ export function isInStatesOrEvents(
 ): boolean {
   const { states } = rule
   // If no states are specified, then the rule applies to all VehicleStates.
-  if (states === null || states === undefined) {
-    return true
-  }
+  if (states === null || states === undefined) return true
 
-  /**
-   * State encoded in the event payload (default acc in the reducer) + states that it is
-   * possible to transition into with any transient event_types
+  const { vehicle_state } = event
+  /*
+   * Have to use a let statement to capture the result of the checks.
+   * Returning the boolean from .forEach means returning from the callback function, not
+   * .isInStatesOrEvents
    */
-  const possibleStates = getPossibleStates(device, event)
-
-  return possibleStates.some(state => {
-    // Explicit events encoded in rule for that state (if any)
-    // eslint-disable-next-line @typescript-eslint/no-explicit-any
-    const matchableEvents: string[] | undefined = state in states ? (states as any)[state] : undefined //FIXME should not have to use an any cast here
-
-    /**
-     * If event_types not encoded in rule, assume state match.
-     * If event_types encoded in rule, see if event.event_types contains a match.
-     */
-    if (
-      matchableEvents !== undefined &&
-      (matchableEvents.length === 0 || areThereCommonElements(matchableEvents, event.event_types))
-    ) {
-      return true
+  let result = false
+  Object.keys(states).forEach(state => {
+    if (state === vehicle_state) {
+      // FIXME It might be possible to avoid this `any` cast...
+      // eslint-disable-next-line @typescript-eslint/no-explicit-any
+      const ruleEventTypes = (states as any)[vehicle_state]
+      if (!ruleEventTypes || ruleEventTypes.length === 0) {
+        result = true
+        return
+      }
+      if (areThereCommonElements(event.event_types, ruleEventTypes)) {
+        result = true
+        return
+      }
```

Comment:
I think we might be able to avoid iterating through all of the `states` here and just do a lookup in the map instead!

Something kinda like:
```
const ruleEventTypes = states[vehicle_state] // I think this will be VEHICLE_EVENT[] | undefined

if (!ruleEventTypes || ruleEventTypes.length === 0) return true

return areThereCommonElements(event.event_types, ruleEventTypes)
```

Reply:
Don't you now have to iterate through all the possible vehicle states, i.e. iterate through `['available', 'reserved'...]` instead now?

Reply:
Oh d'oh I see what you mean, just tired